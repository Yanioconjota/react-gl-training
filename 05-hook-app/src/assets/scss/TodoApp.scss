$todoBg: #2a183c;
$fontColor: #fff;

.list-group {
  .list-group-item {
    background-color: lighten($color: $todoBg, $amount: 10%);
    color: $fontColor;
    padding: 0;
    margin-bottom: 7px;
    p {
      cursor: pointer;
      margin: 0;
      padding: 7px 0 7px 12px;
      width: 100%;
      &:hover {
        background-color: lighten($color: $todoBg, $amount: 20%);
      }
    }
  }
}

.complete {
  text-decoration: line-through;
  color: darken($color: $fontColor, $amount: 35%);
}