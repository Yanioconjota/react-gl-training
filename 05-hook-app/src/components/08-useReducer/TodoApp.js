import React, { useReducer } from 'react';
import {Row, Col, Button, ButtonGroup} from 'react-bootstrap';
import { todoReducer } from './todoReducre';
import '../../assets/scss/TodoApp.scss';

const initialState = [{
  id: new Date().getTime(),
  desc: 'Aprender React',
  done: false
}];

const TodoApp = () => {

  const [ todos ] = useReducer(todoReducer, initialState);

  console.log(todos);

  return (
    <>
      <h1>
        TodoApp <small>({todos.length})</small>
      </h1>
      <hr />
      <Row>
        <Col md={7}>
          <ol className="list-group list-group-flush">
            {todos.map((todo, i) => (
              <li
                key={todo.id}
                className="list-group-item d-flex justify-content-between align-items-center"
              >
                <p className="m-0">
                  {i + 1}. {todo.desc}
                </p>
                <ButtonGroup>
                  <Button variant="danger">Borrar</Button>
                </ButtonGroup>
              </li>
            ))}
          </ol>
        </Col>
        <Col md={5}>
          <h4>Agregar todo</h4>
          <hr />
          <form>
            <input
              className="form-control"
              type="text"
              name="description"
              autocompplete="off"
              placeholder="Aprender..."
            />
            <div className="d-grid mt-2">
              <Button variant="outline-primary">Agregar</Button>
            </div>
          </form>
        </Col>
      </Row>
    </>
  );
};

export default TodoApp;
