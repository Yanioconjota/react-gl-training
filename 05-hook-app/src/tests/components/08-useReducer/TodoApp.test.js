import "@testing-library/jest-dom";
import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";

describe('pruebas con TodoApp', () => {
  const wrapper = shallow(<TodoApp/>);
  //Mock del guardado de los items en local storage
  Storage.prototype.setItem = jest.fn(()=>{});
  test('debe mostrar el componente correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe agregar un todo', () => {
    //se usa mount para simular la ejecución del handleAddTodo
    const wrapper = mount(<TodoApp/>);
    act(()=>{
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
    });
    console.log(wrapper.find('h1').text());
    expect(wrapper.find('h1').text()).toBe('TodoApp (2)');
    //Se espera que se llame por cada todo
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    //emula la ejecución del setItem en cada useEffect
    expect(localStorage.setItem).toHaveBeenCalledWith(
      'todos',
      JSON.stringify(demoTodos)
    );
  });
  
  test("debe eliminar un todo", () => {
    act(() => {
      //Como es un mock debo agregar uno y borrarlo para que funcione la prueba
      wrapper.find("TodoAdd").prop("handleAddTodo")(demoTodos[0]);
      wrapper.find("TodoList").prop("handleDelete")(demoTodos[0].id);
    });
    expect(wrapper.find("h1").text()).toBe("TodoApp (0)");
  });
});
