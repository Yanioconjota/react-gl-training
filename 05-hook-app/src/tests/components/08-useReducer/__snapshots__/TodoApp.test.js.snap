// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas con TodoApp debe mostrar el componente correctamente 1`] = `
<Fragment>
  <h1>
    TodoApp 
    <small>
      (
      0
      )
    </small>
  </h1>
  <hr />
  <Row>
    <Col
      md={7}
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </Col>
    <Col
      md={5}
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </Col>
  </Row>
</Fragment>
`;
